### PR TITLE
Add suport for CHDK created raw files

### PR DESCRIPTION
**Pull request by pedrocr**
Someone thought it was a good idea to create raw files that are just the raw data without anyting else. dcraw identifies these by the exact file size and we do the same here. This also adds a 16bit version of the MSB pump, which is needed for these odd packed 10bit files.

### CODE REVIEW

**Comment by klauspost** [comment]
What is the reason for the BlackArea change?

File: RawSpeed/BitPumpMSB16.cpp [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    You should have received a copy of the GNU Lesser General Public
+    License along with this library; if not, write to the Free Software
+    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+
+    http://www.klauspost.com
+*/
+
+namespace RawSpeed {
+
+/*** Used for entropy encoded sections, for now only Nikon Coolpix ***/
```
Comment (klauspost) [inline RawSpeed/BitPumpMSB16.cpp:27]: Copy&Paste? (Is it CoolPix)

Comment (pedrocr) [inline RawSpeed/BitPumpMSB16.cpp:27]: Yeah, I just copied another BitPump and made changes

**Comment by pedrocr** [comment]
> What is the reason for the BlackArea change?

To make it easier to specify black areas that are on the right or bottom edges. That way a 30pix black area on the right is just -30,30 instead of having to put width-30,30. It's equivalent to what is already done in crops.

**Comment by klauspost** [comment]
Ah - it's relative to opposite side. Didn't realize that.

It makes sense, but I am hesitant to make an incompatible change to the XML format, so it cannot be read by earlier versions (and other stuff that relies on current behaviour).

I will also see if I can find a good way to avoid having the hardcoded table.

**Comment by pedrocr** [comment]
The hardcoded table is needed to at least do the filesize->(make,model) mapping, but maybe there's a way to lookup cameras.xml by a new filesize hint?

**Comment by klauspost** [comment]
Would this make sense: I made a [translated table](https://gist.github.com/klauspost/55f18c5256965c9e1cc5) with a small app.

The difficult/annoying part is that the RawParser will need the CameraMetadata object to be able to identify them.

**Comment by pedrocr** [comment]
> Would this make sense: I made a translated table with a small app.

This looks fine but the white point is wrong. The relevant dcraw code is:

``` C
maximum = (1 << tiff_bps) - (1 << table[i].max);
```

table.max is what you've put in the white point. tiff_bps is calculated from file size, width and height. This is also missing hints for width and height.

> The difficult/annoying part is that the RawParser will need the CameraMetadata object to be able to identify them.

Yep, that's the issue with doing it this way. Haven't looked at the plumbing needed to get that data to that point.

**Comment by klauspost** [comment]
I updated the xml and added PR #85 with a unobtrusive proposal. Does it make sense to you?

**Comment by pedrocr** [comment]
This looks fine to me but it changes the rawspeed API. If you're ok with that the darktable changes look quite small so it wouldn't be an issue.

**Comment by klauspost** [comment]
It is actually compatible, the `virtual RawDecoder* getDecoder(CameraMetaData* meta = NULL);` makes it so that NULL is a default value, if none is specified. Therefore un-updated code will not have chdk, but otherwise still work.

Should I merge #85?  Would you like to adjust this, or should I do the adjustments?

**Comment by pedrocr** [comment]
#85 looks great. Your call on this one. I'm happy to adjust this myself if it helps.

**Comment by pedrocr** [comment]
Should I rework this then?

**Comment by klauspost** [comment]
Cool - I have merged it. 

I would prefer not to include the negative blackwindow sizes for now. I can see how it makes things easier, but on the other hand I don't see enough benefit to break the format.

**Comment by klauspost** [comment]
Yes - please do. I still have small and a big leak to hunt down from todays merges, so I will do that, and probably head to bed soon.

**Comment by pedrocr** [comment]
This should be ready to merge now. Works fine in darktable

**Comment by klauspost** [comment]
Looks very good!

**Pull request merged**